### PR TITLE
feat: support exporting lists as markdown

### LINKED ISSUE #52: Support Exporting to Markdown
The user should be able to export a list to markdown so that it can be edited/viewed in another app (like [obsidian](https://obsidian.md)).

Up for discussion: should the user be able to import a markdown file as a list?

### PR DESCRIPTION
**Pull request by tycebrown**
## What has changed
Added a menu button which, when clicked, allows the user to save their list as a markdown file

## Connected Issues
Closes #52 

## QA Steps
1. Open a list.
2. There should be an item in the "more" popup menu (at the top right) which, when clicked, brings up a dialog to export the current list to markdown. 
3. When the user finishes exporting the list to markdown, there should be a .md file that:
   - has the name the user gave it
   - should contain a reasonable representation of the list in markdown.

### CODE REVIEW

**Comment by tycebrown** [comment]
~~At the current moment, it does not work on android and, in theory, doesn't work on iOS.~~

**Comment by tycebrown** [comment]
## TODO:
- [x] Make the "Export as Markdown" button show a dialog on Desktop. 
- [x] Allow the user to select whether they want the lists labels (tags) in the markdown
  - ~~At the current moment, labels are not put into the markdown.~~
- [ ] ~~?Allow the user to be able to preview the markdown before it is saved to a file?~~ 
- [x] Automatically select and focus on file name in mobile

**Comment by tycebrown** [comment]
Ready for review!

File: lib/view/export_as_markdown_mobile_dialog.dart [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  Widget build(BuildContext context) {
+    return AlertDialog(
+      content: Column(
+        mainAxisSize: MainAxisSize.min,
+        children: [
+          if (widget.includeFileNameTextField)
+            TextField(controller: controller, autofocus: true),
+          const SizedBox(height: 16),
+          Row(mainAxisAlignment: MainAxisAlignment.spaceBetween, children: [
+            const Text('Include Labels?'),
```
Comment (Pertempto) [inline lib/view/export_as_markdown_mobile_dialog.dart]:
```suggestion
            const Text('Include Labels'),
```
It's not really common practice to use question marks in checkbox labels

File: lib/view/list_widget.dart [hunk 2]
```diff
@@ -139,6 +152,54 @@ class _ListWidgetState extends State<ListWidget> {
...
 
+  Future<void> _exportAsMarkdown() async {
+    // We have to have separate exportAsMarkdown functions for mobile and desktop
+    // because the file_picker package (https://pub.dev/packages/file_picker) only
+    // supports saving files in desktop. This package seemed far superior to its competitors.
+    if (Platform.isAndroid || Platform.isIOS) {
+      await _exportAsMarkdownOnMobile();
+    } else {
+      await _exportAsMarkdownOnDesktop();
+    }
```
Comment (Pertempto) [inline lib/view/list_widget.dart]: The list widget really shouldn't need to worry about the platform differences. I would think all this logic should be contained inside the `ExportListAsMarkdownDialog` code

File: lib/view/export_as_markdown_mobile_dialog.dart (hunk 1 above)
Comment (tycebrown) [reply lib/view/export_as_markdown_mobile_dialog.dart]: Done

File: lib/view/list_widget.dart (hunk 2 above)
Comment (tycebrown) [reply lib/view/list_widget.dart]: Done

**Comment by tycebrown** [comment]
One question: At the current moment, repeat info is not stored in the markdown. Should it?

**Comment by Pertempto** [comment]
> One question: At the current moment, repeat info is not stored in the markdown. Should it?

I think it's fine to leave it out.

File: lib/view/list_widget.dart [hunk 3]
```diff
@@ -214,6 +220,21 @@ class _ListWidgetState extends State<ListWidget> {
...
 
+  Future<void> _exportAsMarkdown() async {
+    await showDialog(
+        context: context,
+        builder: (context) => ExportListAsMarkdownDialog(
+            listModel: listModel,
+            onSuccessfulExport: _onSuccessfulMarkdownExport));
+  }
+
+  void _onSuccessfulMarkdownExport() {
```
Comment (Pertempto) [inline lib/view/list_widget.dart]: Why not have this inside the dialog too? Can you not call `ScaffoldMessenger.of(context).showSnackBar()` from there?

**Comment by Pertempto** [comment]
When I run on my Android phone, it doesn't give me any option to select a folder to save the file. It just saves it to my Downloads folder. Is that normal?

File: lib/view/list_widget.dart (hunk 3 above)
Comment (tycebrown) [reply lib/view/list_widget.dart]: Oh, that makes sense to move it. Done.

**Comment by tycebrown** [comment]
> When I run on my Android phone, it doesn't give me any option to select a folder to save the file. It just saves it to my Downloads folder. Is that normal?

That's what I had originally settled on, but it certainly makes more sense that the user should be able to choose a directory to save their file into. Changing...

**Comment by tycebrown** [comment]
Ready for review! I think after this PR, I'll pause and head over to React. Do you think this is a good decision?

**Comment by Pertempto** [comment]
> Ready for review! I think after this PR, I'll pause and head over to React. Do you think this is a good decision? 

Nope. I think it's a GREAT decision! It's always good to broaden your knowledge base. I think you are pretty good at Flutter now.

**Comment by Pertempto** [comment]
The markdown generation works great!

**Comment by Pertempto** [comment]
Your app doesn't display any error message when I try to save a file to my SD card.

I see this error in the logs:

```
Unhandled Exception: PathAccessException: Cannot open file, path = '/storage/394A-F45D/ObsidianVault/alphabet.md' (OS Error: Operation not permitted, errno = 1)
```

**Review by Pertempto — CHANGES_REQUESTED** [review]
Can you please take a look at my error with saving to the SD card? That is where I keep my Obsidian files, so it's pretty important. Maybe you need additional Android permissions to save to an SD card?

Also, the markdown generation works great!
The one thing I noticed - my custom ordering of items doesn't seem to be saved in the markdown file. Can you make `asMarkdown` use the correct ordering?

**Comment by tycebrown** [comment]
> Can you please take a look at my error with saving to the SD card? That is where I keep my Obsidian files, so it's pretty important. Maybe you need additional Android permissions to save to an SD card?

Will look into it.

> Also, the markdown generation works great! The one thing I noticed - my custom ordering of items doesn't seem to be saved in the markdown file. Can you make `asMarkdown` use the correct ordering?

Will do!

**Comment by tycebrown** [comment]
~Hmm... I haven't been able to reproduce the sdcard saving issue. I made a small tweak in the android manifest file, which might change results on older versions of android.~

~Is it still giving the error?~

**Comment by tycebrown** [comment]
Ah, I was able to reproduce!

**Comment by tycebrown** [comment]
Okay, I was able to get the app to work by requesting ALL permissions (though I lose connection to the app in the process when I run from vscode through a USB cable). Seems to indicate that there is/are some permission(s) that I need to request. Digging...

**Comment by tycebrown** [comment]
~There might also be a problem in the way I am saving files..~
There might be a _better_ way to save files.

**Review by Pertempto — APPROVED** [review]

**Pull request merged**